### DISCUSSION TITLE
PDM constraint "==2.15.4" doesn't work

### DISCUSSION THREAD

**Discussion by deronnax** [renovatebot/renovate#30041] [Request Help] [answered]
### Tell us more.

I can't find it in the docs anymore but we used to use the `constraints` to lock Renovate to a given version of Pipenv, it worked well and was very handy. Now we would need the same things for PDM.

**Comment by viceice** [comment]
https://github.com/renovatebot/renovate/blob/c55dc8e4ec9dd2ba897553291b5c5bc80c7b8c03/lib/modules/manager/pep621/processors/pdm.ts#L112

should work

**Reply by deronnax** [reply]
Hum strange, I will check why it's not working. In the meantime, Thank you

**Reply by deronnax** [reply]
I have this warning:
```
 WARN: Invalid tool constraint (repository=company/mainproduct, branch=renovate/sentry-sdk-2.x)
       "toolName": "pdm",
       "constraint": "==2.15.4",
       "versioning": "semver"
```
Is `==2.15.4` incorrect? (we used the same constraint format for Pipenv).

**Reply by rarkins** [reply, marked as answer]
It's invalid for semver versioning. Try removing the ==

**Reply by viceice** [reply]
https://github.com/renovatebot/renovate/blob/9c0b54b925cd12c150034bd53d9e15a4a7a23a8e/lib/util/exec/containerbase.ts#L134-L138

maybe it should pep440 ?

**Reply by deronnax** [reply]
> It's invalid for semver versioning. Try removing the ==

it workd, thank you. I did not know it was invalid for semver.

> maybe it should pep440 ?

What do you mean?
